crates/telio-dns/src/nameserver.rs

```diff
@@ -64,6 +66,309 @@ impl LocalNameServer {
+                    | TunnResult::WriteToTunnelV6(packet, _) => {
+                        // !
+                        let _lease = match semaphore.try_acquire() {
+                            Ok(lease) => lease,
+                            Err(_) => {
+                                telio_log_debug!("Error semaphore.try_acquire()");
+                                return;
+                            }
+                        };
+
+                        let nameserver = nameserver.clone();
+                        let length = match LocalNameServer::process_packet(
+                            nameserver,
+                            packet,
+                            &mut receiving_buffer,
+                        )
+                        .await
+                        {
+                            Ok(length) => length,
+                            Err(e) => {
+                                telio_log_error!("{}", e);
+                                return;
+                            }
+                        };
+
+                        match peer.encapsulate(
+                            receiving_buffer.get(..length).unwrap_or(&[]),
+                            &mut sending_buffer,
+                        ) {
+                            TunnResult::WriteToNetwork(dns) => {
+                                if let Err(e) = socket.send_to(dns, dst_address).await {
+                                    telio_log_warn!(
+                                        "[DNS] Failed to send DNS query response  {:?}",
+                                        e
+                                    )
+                                };
+                            }
+                            TunnResult::Err(e) => {
+                                telio_log_warn!(
+                                    "[DNS] Failed to encapsulate DNS query response  {:?}",
+                                    e
+                                )
+                            }
+                            _ => {}
+                        }
+                    }
+                    _ => {}
+                }
+            });
+        }
+    }
+
+    async fn process_packet(
+        nameserver: Arc<RwLock<LocalNameServer>>,
+        request_packet: &[u8],
+        response_packet: &mut [u8],
+    ) -> Result<usize, String> {
+        let is_ipv4 = request_packet
+            .first()
+            .ok_or_else(|| String::from("Empty request packet"))
+            .and_then(|first_byte| {
+                if first_byte >> 4 == 4 {
+                    Ok(true)
+                } else if first_byte >> 4 == 6 {
+                    Ok(false)
+                } else {
+                    Err(String::from("Invalid IP version"))
+                }
+            })?;
+
+        let resolver = Resolver::new();
+        let (dns_response, udp_destination, udp_source) = {
+            let payload;
+            let (udp_request, dns_source) = if is_ipv4 {
```

Comment:
This is no what I had in mind. Technically there is now one function but it doesn't solve the code duplication issue. I was hoping for something like [process_inbound_ip_packet](https://github.com/NordSecurity/libtelio/blob/93780b99b588d6cc66b1c1d25e89cb1ae3114905/crates/telio-firewall/src/firewall.rs#L374). This way you will have common code in only one place. This doesn't have to be done at compile time since I doubt dns responses need to be maximally fast but however you chose to do it please make sure that the common parts are not duplicated.